lib/offline/session_deleter.js

```diff
@@ -91,7 +91,13 @@ shaka.offline.SessionDeleter = class {
         shaka.log.warning('Error deleting offline session', e);
       }
     }));
-    await drmEngine.destroy();
+
+    // Destroying the engine with no sessions makes the promise not return.
+    // This will happen if there is a problem removing a session.
+    if (sessionIds.length > 0) {
+      await drmEngine.destroy();
+    }
```

Comment:
Alright. I have the unit tests running on my machine now.  It seems there are multiple situations that the session can be rejected. It does seem to get resolved/rejected on the session deleted, so I moved the potential fix to that file to try to handle the situation.

It seems like calling drmEngine.destroy() is not being resolved or rejected, so that is where it is hanging now.

Right now the fix is to skip it if there are no session IDs to actually remove. I am having a hard time telling what that drmEngine.destroy() is actually doing. Will it create a side effect of issues with how I did it? Maybe there is a better way to catch issues or exceptions in that file

Reply:
It looks like one test failed on the automated build on latest mac firefox. The stack from the error, though, didn't really seem to reference the area in my code, so not sure if that is a flaky test, or a real situation where drmEngine.destroy() has to be done with removing sessions that have no sessions to remove.

Kind of not sure how to proceed with this PR. Maybe it isn't even worth it for the edge case it is.

I could probably do a Promise.race() in my application code. Then abort if the remove() promise doesn't resolve/reject anything after X seconds.

Reply:
> It looks like one test failed on the automated build on latest mac firefox. The stack from the error, though, didn't really seem to reference the area in my code, so not sure if that is a flaky test, or a real situation where drmEngine.destroy() has to be done with removing sessions that have no sessions to remove.

Definitely a known-flaky test.  The infrastructure on GitHub Actions is relatively new, and there are still a few flaky tests we haven't fixed yet.

Reply:
> It seems like calling drmEngine.destroy() is not being resolved or rejected, so that is where it is hanging now.
> 
> Right now the fix is to skip it if there are no session IDs to actually remove. I am having a hard time telling what that drmEngine.destroy() is actually doing. Will it create a side effect of issues with how I did it? Maybe there is a better way to catch issues or exceptions in that file

To me, this implies a bug in DrmEngine.  `destroy()` should always resolve, never reject, and hanging is definitely not correct behavior, regardless of the session count.

Thank you for debugging this so deeply!

If you're willing to look a little deeper, can you see if this line in DrmEngine's `destroyNow_()` is the cause of `destroy()` hanging?

> ```js
>     await this.closeOpenSessions_();
> ```

That private method will await on Promise.all, on an array mapped from a list of open sessions:

> ```js
>     await Promise.all(openSessions.map(async ([session, metadata]) => {
> ```

So it would also be interesting to know what size/contents of `openSessions`.  If it's empty, `await Promise.all([])` would be an instant success, so I doubt that it is empty.  Assuming it's not empty, I'd also love to know if you see execution of the code path for `session.remove()` or `this.closeSession_(session)` within that map.

This feels like a Chrome or Widevine bug, and we have a `Promise.race`-based timeout to workaround one of those already.  (See `closeSession_` and http://crbug.com/1108158)